loaders/src/glTF/2.0/Extensions/EXT_mesh_gpu_instancing.ts

```diff
@@ -54,9 +59,19 @@ export class EXT_mesh_gpu_instancing implements IGLTFLoaderExtension {
                 return promise;
             }
 
+            let useThinInstancesForAllMeshes = true;
+            let canUseThinInstances = false;
+
             // Hide the source meshes.
             for (const babylonMesh of node._primitiveBabylonMeshes) {
-                babylonMesh.isVisible = false;
+                if (!(babylonMesh as Mesh).thinInstanceSetBuffer) {
```

Comment:
Chatted with @sebavan. We shouldn't check for existence of this function since if someone is loading this glTF extension, we want to use thin instances. Import the thinInstanceMesh at the top of this file to load it and then you shouldn't have to branch for this. It should simplify all the changes in this PR quite a bit?

Reply:
The problem comes from the fact that as an optimization, the glTF loader creates an `InstancedMesh` instead of a `Mesh` when possible (see line 784 in **glTFLoader**). But if this mesh has "gpu instances" (as per the extension), creating thin instances for it won't work because `InstancedMesh` don't support thin instances.

The easiest way to deal with that would be to disable this optimization completely when **EXT_mesh_gpu_instancing** is used, but we don't know in advance if it will be used or not...

So that would require pre-parsing the file to look for the extension usage. Or is there a faster way to know? I think a glTF file can be flagged to say "this extension XXX is used/required", maybe we can rely on that?

In that case, I can indeed remove all my ugly code :)

Reply:
I see. I suppose it is possible for a glTF to have both, but yeah, I agree if this extension exists then it is unlikely someone wanted both mesh instancing and instancing with buffers. I think ideally we should be able to disable calling createInstance if this extension is in use for this node. I just saw that you added a flag to disable creation of instances. Instead of doing that globally, maybe you can do it per mesh?

Reply:
> I just saw that you added a flag to disable creation of instances. Instead of doing that globally, maybe you can do it per node?

How would that work?

Say I have two nodes, each pointing to the same mesh, and only the second node has the gpu instancing extension.

So, when handling the second node, an `InstancedMesh` will be created because the mesh from the first node was not flagged as "don't allow instancing". We would have to check for the gpu instancing extension right in the glTF loader, when deciding to create an `InstancedMesh` or not, but I don't think we want to do that?

Reply:
This extension is on a glTF node, which is before the mesh is loaded. I think you can change your _disableInstancedMesh flag into a ref count and increment the ref count right before calling the original loadNode: https://github.com/BabylonJS/Babylon.js/pull/8213/files#diff-3dddcf4f449da13644db39f2e82479f4R52. This should cause the core loader to not create an instanced mesh. Then decrement the ref count when the extension is done loading.

Reply:
Hmm, though this may cause an issue with child nodes that should create an instanced mesh. I wonder if the spec should allow a node with this extension to have child nodes. 🤔

Reply:
It seems something like this work:
```typescript
public loadNodeAsync(context: string, node: INode, assign: (babylonTransformNode: TransformNode) => void): Nullable<Promise<TransformNode>> {
    return GLTFLoader.LoadExtensionAsync<IEXTMeshGpuInstancing, TransformNode>(context, node, this.name, (extensionContext, extension) => {
        this._loader._disableInstancedMesh = true;

        const promise = this._loader.loadNodeAsync(`#/nodes/${node.index}`, node, assign);

        this._loader._disableInstancedMesh = false;
```
instead of doing a global disabling.

Would that work (must say I don't understand how the ref count would make the core loader not to use instanced meshes without making changes to the loader itself)?

Reply:
Yes, this should work if there are no children. I was suggesting ref counting since there maybe a recursive load because loadNodeAsync traverses the nodes.

Reply:
Something like:
```typescript
public loadNodeAsync(context: string, node: INode, assign: (babylonTransformNode: TransformNode) => void): Nullable<Promise<TransformNode>> {
    return GLTFLoader.LoadExtensionAsync<IEXTMeshGpuInstancing, TransformNode>(context, node, this.name, (extensionContext, extension) => {
        this._loader._disableInstancedMesh++;

        const promise = this._loader.loadNodeAsync(`#/nodes/${node.index}`, node, assign);

        this._loader._disableInstancedMesh--;
```
and testing for `this._loader._disableInstancedMesh === 0` on the loader side?

Sorry for all those back and forth, but I have little clue how everything works under the hood in the gltf loader (and that does not seem easy to grasp with all the async jobs going on).

Reply:
Yup, exactly.

> Sorry for all those back and forth

No need to be. Thanks for all the work!